go/vt/vttablet/tabletserver/repltracker/writer_test.go

```diff
@@ -47,16 +48,222 @@ func TestWriteHeartbeat(t *testing.T) {
+		tw.RequestHeartbeats()
+		ctx, cancel := context.WithTimeout(context.Background(), tw.onDemandDuration-time.Second)
+		defer cancel()
+		ticker := time.NewTicker(1 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				assert.EqualValues(t, 0, writeErrors.Get())
+				currentWrites := writes.Get()
+				assert.Greater(t, currentWrites, lastWrites)
+				lastWrites = currentWrites
+			}
+		}
+	})
+	t.Run("exhaust lease, no heartbeats", func(t *testing.T) {
+		<-time.After(tw.onDemandDuration)
+		currentWrites := writes.Get()
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, currentWrites, writes.Get())
+	})
+	tw.Close()
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.Nil(t, rateLimiter)
+	}
+}
+
+// TestWriteHeartbeatOnDemand tests that the heartbeat writer initiates leased heartbeats once opened,
+// and then upon RequestHeartbeats().
+func TestWriteHeartbeatOnDemand(t *testing.T) {
+	defaultOnDemandDuration = 7 * time.Second
+	onDemandDuration := 3 * time.Second
+
+	db := fakesqldb.New(t)
+	defer db.Close()
+
+	tw := newTestWriter(db, nil, tabletenv.Heartbeat, onDemandDuration)
+
+	assert.Equal(t, onDemandDuration, tw.onDemandDuration)
+
+	db.AddQueryPattern("^INSERT INTO.*", &sqltypes.Result{})
+
+	writes.Reset()
+	writeErrors.Reset()
+
+	tw.writeHeartbeat()
+	lastWrites := writes.Get()
+	assert.EqualValues(t, 1, lastWrites)
+	assert.EqualValues(t, 0, writeErrors.Get())
+
+	t.Run("closed, no heartbeats", func(t *testing.T) {
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, 1, writes.Get())
+	})
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.Nil(t, rateLimiter)
+	}
+	tw.Open()
+	defer tw.Close()
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.NotNil(t, rateLimiter)
+	}
+	t.Run("open, initial heartbeats", func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(context.Background(), tw.onDemandDuration-time.Second)
+		defer cancel()
+		ticker := time.NewTicker(1 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				assert.EqualValues(t, 0, writeErrors.Get())
+				currentWrites := writes.Get()
+				assert.Greater(t, currentWrites, lastWrites)
+				lastWrites = currentWrites
+			}
+		}
+	})
+	t.Run("exhaust lease, no heartbeats", func(t *testing.T) {
+		<-time.After(tw.onDemandDuration)
+		currentWrites := writes.Get()
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, currentWrites, writes.Get())
+	})
+	t.Run("request heartbeats, heartbeats", func(t *testing.T) {
+		tw.RequestHeartbeats()
+		lastWrites := writes.Get()
+		<-time.After(tw.onDemandDuration)
+		assert.Greater(t, writes.Get(), lastWrites)
+	})
+	t.Run("exhaust lease, no heartbeats", func(t *testing.T) {
+		<-time.After(tw.onDemandDuration)
+		currentWrites := writes.Get()
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, currentWrites, writes.Get())
+	})
```

Comment:
This seems redundant to the one above.

Reply:
It is not. The first one validates that the heartbeat lease is exhausted just after `Open()` is called. The 2nd validates that heartbeats lease is exhausted after `RequestHeartbeats()` is called.